stores.js

```diff
@@ -0,0 +1,260 @@
+      transform (result, enc, next) {
+        if (result.value === null) {
+          next()
+          return
+        }
+        const { value, sig, seq: storedSeq } = result.value
+        const msg = salt ? Buffer.concat([value, salt]) : value
+        if (storedSeq >= userSeq && verify(sig, msg, key)) {
+          if (streamMode) {
+            next(null, { value, sig, seq: storedSeq, salt })
+          } else {
+            found = true
+            cb(null, { value, sig, seq: storedSeq, salt })
+            queryStream.destroy()
+          }
+        }
+      }
+    })
+    finished(queryStream, () => { responseStream.push(null) })
+    pipeline(queryStream, responseStream, (err) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      if (streamMode === false && found === false) {
+        cb(null, null)
+      }
+    })
+    if (streamMode) return responseStream
+  }
+  put (value, opts, cb) {
+    if (!Buffer.isBuffer(value)) throw Error('Value must be a buffer')
+    if (value.length > PUT_VALUE_MAX_SIZE) { throw Error(`Value size must be <= ${PUT_VALUE_MAX_SIZE}`) }
+    if (typeof opts !== 'object') throw Error('Options are required')
+    if (typeof cb !== 'function') throw Error('Callback is required')
+    if (value.length > PUT_VALUE_MAX_SIZE) { throw Error(`Value size must be <= ${PUT_VALUE_MAX_SIZE}`) }
+    const { dht } = this
+    const { seq = 0, salt, keypair } = opts
+    if (!keypair) throw Error('keypair is required')
+    const { secretKey, publicKey } = keypair
+    if (!Buffer.isBuffer(secretKey)) throw Error('keypair.secretKey is required')
+    if (!Buffer.isBuffer(publicKey)) throw Error('keypair.publicKey is required')
+    if (typeof seq !== 'number') throw Error('seq should be a number')
+    if (salt) {
+      if (!Buffer.isBuffer(salt)) throw Error('salt must be a buffer')
+      if (salt.length < 16 && salt.length > 64) { throw Error('salt size must be between 16 and 64 bytes (inclusive)') }
+    }
+    const sig = Buffer.alloc(signSize)
+    const msg = salt ? Buffer.concat([value, salt]) : value
+    sign(sig, msg, secretKey)
+    const key = publicKey
+    const info = { value, sig, seq, salt }
+
+    dht.update('mutable-store', key, info, (err) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      cb(null, key, info)
+    })
+  }
+  _command () {
+    const { store } = this
+    return {
+      valueEncoding: mutableEncoding,
+      update (input, cb) {
+        if (input.value == null) {
+          cb(null)
+          return
+        }
+        const publicKey = input.target
+        const { value, salt, sig, seq } = input.value
+        const key = salt
+          ? publicKey.toString('hex') + salt.toString('hex')
+          : publicKey.toString('hex')
+        const local = store.get(key)
+        const msg = salt ? Buffer.concat([value, salt]) : value
+        const verified = verify(sig, msg, publicKey) &&
+          (local ? seq >= local.seq : true)
+
```

Comment:
okay small detail i forgot to mention ... if seq === local.seq AND value === local.value (and the sig is correct obvs) then simply refresh the LRU (ie not an error). this happens all the time in practice as nodes will republish data to make sure it stays alive.

Reply:
I think this behaviour is already there..

if seq === local.seq will pass seq >= local.seq - so if the sig verifies and seq === local.seq, then `verified` will be true, so `store.set(key, { value, salt, sig, seq })` will execute which will refresh the LRU by re-setting the value. There's no need to check that values equal for this case because the behaviour would be the same whether it's changed or not

Reply:
clarified offline, now I understand